File: ros2_packages/lapgym_ros2_bridge/lapgym_ros2_bridge/retract_policy_server.py
"""
Phase 4B: RetractPolicyServer -- ROS 2 action server wrapping Phase 2D PPO.

Action: lapgym_interfaces/action/Retract
  Goal:     max_steps (float32), render (bool)
  Feedback: distance_to_goal, distance_mm, step, in_collision, collision_cost
  Result:   success, steps_taken, final_distance, termination

The server loads the Phase 2D PPO checkpoint and runs policy.predict(obs)
every step. It checks is_preempted() after every step so the Behaviour Tree
(Phase 4C) can cancel the goal at any time safely.

Author: Subhash Arockiadoss
"""

# -- Gymnasium shim MUST be first ---------------------------------------------
import sys
import gymnasium
sys.modules['gym'] = gymnasium
sys.modules['gym.spaces'] = gymnasium.spaces
# -----------------------------------------------------------------------------

import os
import threading
import numpy as np
import rclpy
from std_msgs.msg import Bool
from rclpy.node import Node
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup

from lapgym_interfaces.action import Retract

# PPO checkpoint path relative to surgical-rl repo root
CHECKPOINT_PATH = 'logs/checkpoints/phase2_ppo_tissue_retraction_20260409_211946/ppo_tissue_final'
DEFAULT_MAX_STEPS = 300


class RetractPolicyServer(Node):
    """ROS 2 action server that runs the Phase 2D PPO policy."""

    def __init__(self):
        super().__init__('retract_policy_server')

        # -- Load PPO policy --------------------------------------------------
        self._policy = None
        self._env = None
        self._load_policy()

        # -- Action server ----------------------------------------------------
        self._action_server = ActionServer(
            self,
            Retract,
            'retract_policy',
            execute_callback=self._execute_cb,
            goal_callback=self._goal_cb,
            cancel_callback=self._cancel_cb,
            callback_group=ReentrantCallbackGroup(),
        )

        # # Surgeon stop -- dedicated background node + thread
        # # env.step() blocks the main executor so we need a separate thread
        self._surgeon_stopped = False
        self._stop_event = threading.Event()
        self._stop_event.set()
        self._emergency = False

        self._stop_context = rclpy.Context()
        self._stop_context.init()

        self._stop_node = rclpy.create_node(
            '_surgeon_stop_retract',
            context=self._stop_context,
            enable_rosout=False
        )
        self._stop_node.create_subscription(
            Bool, '/surgeon_stop', self._surgeon_cb, 10)
        self._stop_node.create_subscription(
            Bool, '/emergency_stop', self._emergency_cb, 10)

        self._stop_executor = rclpy.executors.SingleThreadedExecutor(
            context=self._stop_context)
        self._stop_executor.add_node(self._stop_node)

        self._stop_thread = threading.Thread(
            target=self._spin_stop_node, daemon=True)
        self._stop_thread.start()
        
        self.get_logger().info('RetractPolicyServer ready')


    def _spin_stop_node(self):
        while self._stop_context.ok():
            try:
                self._stop_executor.spin_once(timeout_sec=0.01)
            except Exception:
                pass


    # -- Policy loading -------------------------------------------------------

    def _load_policy(self):
        """Load Phase 2D PPO checkpoint and create environment."""
        try:
            from stable_baselines3 import PPO
            from sofa_env.scenes.tissue_retraction.tissue_retraction_env import RenderMode
            from envs.tissue_retraction_v2 import TissueRetractionV2

            self.get_logger().info(f'Loading PPO checkpoint: {CHECKPOINT_PATH}')
            self._policy = PPO.load(CHECKPOINT_PATH)
            self.get_logger().info('PPO checkpoint loaded successfully')

            # Create environment in headless mode by default
            self._env = TissueRetractionV2(
                env_kwargs={'render_mode': RenderMode.HUMAN})
            obs, _ = self._env.reset()
            self._obs = obs
            self.get_logger().info('TissueRetractionV2 environment ready')

        except Exception as e:
            self.get_logger().error(f'Failed to load policy: {e}')
            self._policy = None
            self._env = None

    # -- Action callbacks -----------------------------------------------------

    def _goal_cb(self, goal_request):
        """Accept or reject incoming goal."""
        self.get_logger().info(
            f'Received goal: max_steps={goal_request.max_steps} '
            f'render={goal_request.render}')
        if self._policy is None or self._env is None:
            self.get_logger().error('Policy not loaded -- rejecting goal')
            return GoalResponse.REJECT
        return GoalResponse.ACCEPT

    def _cancel_cb(self, goal_handle):
        """Accept cancellation requests -- always allow preemption."""
        self.get_logger().info('Cancel request received -- accepting')
        return CancelResponse.ACCEPT
    
    def _surgeon_cb(self, msg: Bool):
        if msg.data:
            self._surgeon_stopped = True
            self._stop_event.clear()
            self.get_logger().warn('Retract: SURGEON STOP received')
        else:
            self._surgeon_stopped = False
            self._stop_event.set()
            self.get_logger().info('Retract: SURGEON RESUME received')

    def _emergency_cb(self, msg: Bool):
        if msg.data and not self._emergency:
            self._emergency = True
            self.get_logger().error('Retract: EMERGENCY STOP received')

    async def _execute_cb(self, goal_handle):
        """Main policy execution loop."""
        self.get_logger().info('Executing retract policy')

        goal = goal_handle.request
        max_steps = int(goal.max_steps) if goal.max_steps > 0 else DEFAULT_MAX_STEPS

        # Reset environment for fresh episode
        obs, _ = self._env.reset()
        self._obs = obs

        feedback_msg = Retract.Feedback()
        step = 0
        termination = 'timeout'
        final_distance = 0.0

        while step < max_steps:

            # -- Check preemption FIRST every step ----------------------------
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                self.get_logger().info(
                    f'Goal preempted at step {step}')
                result = Retract.Result()
                result.success = False
                result.steps_taken = step
                result.final_distance = final_distance
                result.termination = 'preempted'
                return result
            
            # -- Emergency stop check -----------------------------------------
            if self._emergency:
                try:
                    goal_handle.canceled()
                except Exception:
                    goal_handle.abort()
                result = Retract.Result()
                result.success = False
                result.steps_taken = step
                result.final_distance = final_distance
                result.termination = 'emergency_stop'
                return result

            # -- Surgeon stop: freeze until resumed ---------------------------
            while self._surgeon_stopped and not self._emergency:
                if goal_handle.is_cancel_requested:
                    break
                self._stop_event.wait(timeout=0.05)

            # -- Policy inference ---------------------------------------------
            action, _ = self._policy.predict(obs, deterministic=True)

            # -- Step environment ---------------------------------------------
            obs, reward, terminated, truncated, info = self._env.step(action)
            step += 1

            # -- Immediate surgeon stop check after every step ----------------
            while self._surgeon_stopped and not self._emergency:
                if goal_handle.is_cancel_requested:
                    break
                self._stop_event.wait(timeout=0.05)


            # -- Extract guidance info ----------------------------------------
            dist_world = float(
                info.get('distance_to_grasping_position', 0.0) or 0.0)
            in_collision = bool(info.get('in_collision', False))
            collision_cost = abs(
                float(info.get('collision_cost', 0.0) or 0.0))
            final_distance = dist_world

            # -- Feedback every step (for accurate console display) -----------    
            feedback_msg.distance_to_goal = dist_world
            feedback_msg.distance_mm = dist_world * 1000.0
            feedback_msg.step = step
            feedback_msg.in_collision = in_collision
            feedback_msg.collision_cost = collision_cost
            goal_handle.publish_feedback(feedback_msg)
            if step % 5 == 0:
                self.get_logger().info(
                    f'Step {step:3d} | Dist: {dist_world*1000:.1f}mm '
                    f'| {"COL" if in_collision else "SAFE"}')
            # -- Check termination --------------------------------------------
            if terminated:
                goal_reached = bool(info.get('goal_reached', False))
                if goal_reached:
                    termination = 'goal_reached'
                    self.get_logger().info(
                        f'Goal reached at step {step}!')
                else:
                    termination = 'collision'
                    self.get_logger().warn(
                        f'Episode terminated (collision) at step {step}')
                break

        # -- Build result -----------------------------------------------------
        success = termination == 'goal_reached'
        goal_handle.succeed()

        result = Retract.Result()
        result.success = success
        result.steps_taken = step
        result.final_distance = final_distance
        result.termination = termination

        self.get_logger().info(
            f'Episode complete: {termination} | '
            f'steps={step} | dist={final_distance*1000:.1f}mm')
        return result

    # -- Cleanup --------------------------------------------------------------

    def destroy_node(self):
        self._stop_executor.shutdown()
        self._stop_node.destroy_node()
        self._stop_context.shutdown()
        if self._env is not None:
            try:
                self._env.close()
            except Exception:
                pass
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = RetractPolicyServer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
        os._exit(0)

File: ros2_packages/lapgym_ros2_bridge_cpp/src/retract_policy_server.cpp
/**
 * Phase 4G: RetractPolicyServer (C++)
 *
 * ROS 2 action server in C++ that retracts tissue using the Phase 2D
 * PPO policy. Demonstrates the complete hybrid C++/Python pattern:
 *
 *   C++ handles:  action server, threading, stop flags, control loop
 *   Python handles: PPO inference (/ppo_predict), SOFA physics (/sofa_step)
 *
 * Data flow per step:
 *   obs[7] → /ppo_predict → action[3] → /sofa_step → new_obs, dist
 *   repeat until dist < threshold or terminated
 *
 * Author: Subhash Arockiadoss
 */

#include <chrono>
#include <memory>
#include <thread>
#include <atomic>
#include <array>
#include <string>
#include <functional>
#include <future>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/bool.hpp"

#include "lapgym_interfaces/action/retract.hpp"
#include "lapgym_interfaces/srv/sofa_step.hpp"
#include "lapgym_interfaces/srv/ppo_predict.hpp"

static constexpr float GOAL_THRESHOLD  = 0.03f;   // 30mm
static constexpr int   DEFAULT_MAX_STEPS = 300;

using Retract           = lapgym_interfaces::action::Retract;
using SofaStep          = lapgym_interfaces::srv::SofaStep;
using PPOPredict        = lapgym_interfaces::srv::PPOPredict;
using GoalHandleRetract = rclcpp_action::ServerGoalHandle<Retract>;


class RetractPolicyServerCpp : public rclcpp::Node
{
public:
    RetractPolicyServerCpp()
    : Node("retract_policy_server_cpp"),
      surgeon_stopped_(false),
      emergency_(false)
    {
        // -- Action server -------------------------------------------------
        action_server_ = rclcpp_action::create_server<Retract>(
            this,
            "retract_policy_cpp",
            std::bind(&RetractPolicyServerCpp::handleGoal,     this,
                      std::placeholders::_1, std::placeholders::_2),
            std::bind(&RetractPolicyServerCpp::handleCancel,   this,
                      std::placeholders::_1),
            std::bind(&RetractPolicyServerCpp::handleAccepted, this,
                      std::placeholders::_1)
        );

        // -- Service clients — each on dedicated callback group ------------
        // Two services: /sofa_step (physics) and /ppo_predict (ML)
        // Each needs its own callback group so responses can be processed
        // concurrently while execute() loop is running
        sofa_callback_group_ = create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);
        sofa_client_ = create_client<SofaStep>(
            "/sofa_step",
            rmw_qos_profile_services_default,
            sofa_callback_group_);

        ppo_callback_group_ = create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);
        ppo_client_ = create_client<PPOPredict>(
            "/ppo_predict",
            rmw_qos_profile_services_default,
            ppo_callback_group_);

        // Wait for both services
        RCLCPP_INFO(get_logger(), "Waiting for /sofa_step...");
        while (!sofa_client_->wait_for_service(std::chrono::seconds(1))) {
            if (!rclcpp::ok()) return;
        }
        RCLCPP_INFO(get_logger(), "/sofa_step ready");

        RCLCPP_INFO(get_logger(), "Waiting for /ppo_predict...");
        while (!ppo_client_->wait_for_service(std::chrono::seconds(1))) {
            if (!rclcpp::ok()) return;
        }
        RCLCPP_INFO(get_logger(), "/ppo_predict ready");

        // -- Stop listener on dedicated background thread -----------------
        stop_callback_group_ = create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        rclcpp::SubscriptionOptions stop_opts;
        stop_opts.callback_group = stop_callback_group_;

        surgeon_stop_sub_ = create_subscription<std_msgs::msg::Bool>(
            "/surgeon_stop", 10,
            std::bind(&RetractPolicyServerCpp::surgeonStopCb, this,
                      std::placeholders::_1),
            stop_opts);

        emergency_sub_ = create_subscription<std_msgs::msg::Bool>(
            "/emergency_stop", 10,
            std::bind(&RetractPolicyServerCpp::emergencyCb, this,
                      std::placeholders::_1),
            stop_opts);

        stop_executor_ =
            std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
        stop_executor_->add_callback_group(
            stop_callback_group_, get_node_base_interface());
        stop_thread_ = std::thread([this]() {
            stop_executor_->spin();
        });

        RCLCPP_INFO(get_logger(),
            "RetractPolicyServerCpp ready on /retract_policy_cpp");
    }

    ~RetractPolicyServerCpp()
    {
        if (stop_executor_) stop_executor_->cancel();
        if (stop_thread_.joinable()) stop_thread_.join();
    }

private:
    // -----------------------------------------------------------------------
    // Action callbacks
    // -----------------------------------------------------------------------

    rclcpp_action::GoalResponse handleGoal(
        const rclcpp_action::GoalUUID & /*uuid*/,
        std::shared_ptr<const Retract::Goal> /*goal*/)
    {
        RCLCPP_INFO(get_logger(), "Retract goal received — accepting");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handleCancel(
        const std::shared_ptr<GoalHandleRetract> /*goal_handle*/)
    {
        RCLCPP_INFO(get_logger(), "Retract cancel — accepting");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handleAccepted(const std::shared_ptr<GoalHandleRetract> goal_handle)
    {
        std::thread([this, goal_handle]() {
            execute(goal_handle);
        }).detach();
    }

    // -----------------------------------------------------------------------
    // Main execution loop
    // -----------------------------------------------------------------------

    void execute(const std::shared_ptr<GoalHandleRetract> goal_handle)
    {
        RCLCPP_INFO(get_logger(), "Executing retract policy (C++)");

        auto goal     = goal_handle->get_goal();
        int max_steps = (goal->max_steps > 0)
                        ? static_cast<int>(goal->max_steps)
                        : DEFAULT_MAX_STEPS;

        // Reset environment and get initial observation
        auto init = sofaReset();
        // obs is the 7D observation: tool_xyz + goal_xyz + phase
        std::array<float, 7> obs;
        for (int i = 0; i < 7; i++) obs[i] = init->observation[i];

        auto feedback        = std::make_shared<Retract::Feedback>();
        int  step            = 0;
        float final_distance = 0.0f;
        std::string termination = "timeout";

        while (step < max_steps) {

            // -- Preemption check FIRST ------------------------------------
            if (goal_handle->is_canceling()) {
                auto result        = std::make_shared<Retract::Result>();
                result->success        = false;
                result->steps_taken    = step;
                result->final_distance = final_distance;
                result->termination    = "preempted";
                goal_handle->canceled(result);
                return;
            }

            // -- Emergency stop -------------------------------------------
            if (emergency_.load()) {
                auto result            = std::make_shared<Retract::Result>();
                result->success        = false;
                result->steps_taken    = step;
                result->final_distance = final_distance;
                result->termination    = "emergency_stop";
                try { goal_handle->abort(result); } catch (...) {}
                RCLCPP_ERROR(get_logger(), "EMERGENCY STOP at step %d", step);
                return;
            }

            // -- Surgeon stop freeze loop ----------------------------------
            while (surgeon_stopped_.load() && !emergency_.load()) {
                if (goal_handle->is_canceling()) break;
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }

            // -- Step 1: Send obs to PPO, get action ----------------------
            // Python PPOPredictService calls policy.predict(obs)
            // PyTorch releases GIL → GPU computes → returns action[3]
            auto ppo_response = ppoPredict(obs);
            if (!ppo_response->success) {
                RCLCPP_ERROR(get_logger(), "PPO predict failed at step %d", step);
                break;
            }
            std::array<float, 3> action = {
                ppo_response->action[0],
                ppo_response->action[1],
                ppo_response->action[2]
            };

            // -- Step 2: Send action to SOFA, get new state ---------------
            auto sofa_response = sofaStep(action);
            step++;

            // Extract new observation for next PPO call
            for (int i = 0; i < 7; i++) obs[i] = sofa_response->observation[i];

            // Use distance from sofa response
            float dist = sofa_response->distance_to_end_position;
            final_distance = dist;

            // -- Post-step surgeon stop freeze loop -----------------------
            while (surgeon_stopped_.load() && !emergency_.load()) {
                if (goal_handle->is_canceling()) break;
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }

            // -- Check goal reached ---------------------------------------
            if (dist < GOAL_THRESHOLD && dist > 0.0f) {
                termination = "goal_reached";
                RCLCPP_INFO(get_logger(),
                    "Goal reached at step %d dist=%.1fmm",
                    step, dist * 1000.0f);
                break;
            }

            // -- Check terminated (collision) -----------------------------
            if (sofa_response->terminated) {
                termination = "collision";
                break;
            }

            // -- Publish feedback -----------------------------------------
            feedback->distance_to_goal = dist;
            feedback->distance_mm      = dist * 1000.0f;
            feedback->step             = step;
            feedback->in_collision     = sofa_response->in_collision;
            feedback->collision_cost   = sofa_response->collision_cost;
            goal_handle->publish_feedback(feedback);

            if (step % 5 == 0) {
                RCLCPP_INFO(get_logger(),
                    "Retract (C++) step %3d | Dist: %.1fmm",
                    step, dist * 1000.0f);
            }
        }

        // -- Return result ------------------------------------------------
        bool success = (termination == "goal_reached");
        auto result  = std::make_shared<Retract::Result>();
        result->success        = success;
        result->steps_taken    = step;
        result->final_distance = final_distance;
        result->termination    = termination;
        goal_handle->succeed(result);

        RCLCPP_INFO(get_logger(),
            "Retract complete: %s steps=%d dist=%.1fmm",
            termination.c_str(), step, final_distance * 1000.0f);
    }

    // -----------------------------------------------------------------------
    // Service helpers
    // -----------------------------------------------------------------------

    std::shared_ptr<SofaStep::Response> sofaReset()
    {
        auto request    = std::make_shared<SofaStep::Request>();
        request->action = {0.0f, 0.0f, 0.0f};
        request->reset  = true;
        auto future = sofa_client_->async_send_request(request);
        auto status = future.wait_for(std::chrono::seconds(10));
        if (status != std::future_status::ready) {
            RCLCPP_ERROR(get_logger(), "SOFA reset timed out");
            return std::make_shared<SofaStep::Response>();
        }
        RCLCPP_INFO(get_logger(), "SOFA reset complete");
        return future.get();
    }

    std::shared_ptr<SofaStep::Response> sofaStep(
        std::array<float, 3> action)
    {
        auto request    = std::make_shared<SofaStep::Request>();
        request->action = action;
        request->reset  = false;
        auto future = sofa_client_->async_send_request(request);
        auto status = future.wait_for(std::chrono::seconds(5));
        if (status != std::future_status::ready) {
            RCLCPP_ERROR(get_logger(), "SOFA step timed out");
            return std::make_shared<SofaStep::Response>();
        }
        return future.get();
    }

    std::shared_ptr<PPOPredict::Response> ppoPredict(
        std::array<float, 7> obs)
    {
        auto request = std::make_shared<PPOPredict::Request>();
        for (int i = 0; i < 7; i++) request->observation[i] = obs[i];
        auto future = ppo_client_->async_send_request(request);
        auto status = future.wait_for(std::chrono::seconds(5));
        if (status != std::future_status::ready) {
            RCLCPP_ERROR(get_logger(), "PPO predict timed out");
            return std::make_shared<PPOPredict::Response>();
        }
        return future.get();
    }

    // -----------------------------------------------------------------------
    // Stop callbacks
    // -----------------------------------------------------------------------

    void surgeonStopCb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data) {
            surgeon_stopped_.store(true);
            RCLCPP_WARN(get_logger(), "Retract (C++): SURGEON STOP");
        } else {
            surgeon_stopped_.store(false);
            RCLCPP_INFO(get_logger(), "Retract (C++): SURGEON RESUME");
        }
    }

    void emergencyCb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data && !emergency_.load()) {
            emergency_.store(true);
            RCLCPP_ERROR(get_logger(), "Retract (C++): EMERGENCY STOP");
        }
    }

    // -----------------------------------------------------------------------
    // Member variables
    // -----------------------------------------------------------------------

    rclcpp_action::Server<Retract>::SharedPtr                    action_server_;

    // Two service clients — each with dedicated callback group
    rclcpp::Client<SofaStep>::SharedPtr                          sofa_client_;
    rclcpp::CallbackGroup::SharedPtr                             sofa_callback_group_;
    rclcpp::Client<PPOPredict>::SharedPtr                        ppo_client_;
    rclcpp::CallbackGroup::SharedPtr                             ppo_callback_group_;

    std::atomic<bool>                                            surgeon_stopped_;
    std::atomic<bool>                                            emergency_;

    rclcpp::CallbackGroup::SharedPtr                             stop_callback_group_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         surgeon_stop_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         emergency_sub_;
    std::shared_ptr<rclcpp::executors::SingleThreadedExecutor>   stop_executor_;
    std::thread                                                  stop_thread_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RetractPolicyServerCpp>();

    // Four callback groups need concurrent execution:
    // 1. Action server (default)
    // 2. sofa_callback_group_ — /sofa_step responses
    // 3. ppo_callback_group_  — /ppo_predict responses
    // 4. stop_callback_group_ — surgeon stop (on own thread)
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();

    rclcpp::shutdown();
    return 0;
}